Abort cleanup

jitsni opened the pull request.

jitsni committed:
Cleaning up the resources after ABORT. This includes
- ABORT/RESET on all the necessary streams
- Removing the corresponding correlations

jitsni committed:
Merge remote-tracking branch 'upstream/develop' into abort.cleanup

Conflicts:
	pom.xml

jitsni committed:
Cleaning up the streams when SSLException happens

jitsni committed:
Removing .idea files

jitsni committed:
Removing print statements

jitsni committed:
Adding simple clean up for ABORT/RESET/END
- network events clean up application streams (and vice versa)
- network reply events clean up application reply streams (and vice versa)

jitsni committed:
Cleaning up for ABORT/RESET on the client side

jitsni committed:
Removing the special handling of trucation in SSLException

jitsni committed:
If tlsEngine.isOutboundDone() is done, it sends END to application stream.
Now, it also sends ABORT to application(if application hasn't ended already)

jitsni committed:
Merge remote-tracking branch 'upstream/develop' into abort.cleanup

Conflicts:
	pom.xml

jitsni committed:
Ignoring server.sent.close tests as  write close interferes with
handshake completion

jfallows commented on src/main/java/org/reaktivity/nukleus/tls/internal/stream/ClientStreamFactory.java:
```diff
@@ -240,6 +240,7 @@ private RouteFW wrapRoute(
         private int applicationWindowBytesAdjustment;
         private int applicationWindowFrames;
         private int applicationWindowFramesAdjustment;
+        private boolean applicationEnded;
```
Suggest setting `applicationWindowBytes` to `-1` instead of adding the extra `applicationEnded` field.

jfallows commented on src/main/java/org/reaktivity/nukleus/tls/internal/stream/ClientStreamFactory.java:
```diff
@@ -791,6 +819,9 @@ private void updateNetworkWindow(
         private int applicationWindowFrames;
         private int applicationReplySlot = NO_SLOT;
         private int applicationReplySlotOffset;
+        private boolean networkReplyEnded;
```
Suggest setting `networkReplyWindow` to `-1` instead of adding extra `networkReplyEnded` field.

jfallows committed:
Update nukleus-tls.spec dependency version

jitsni committed:
Using networkWindowBytes and applicationWindowBytes instead of flags to detect
the end of stream

jitsni committed:
Removing .idea files

jfallows commented on src/main/java/org/reaktivity/nukleus/tls/internal/stream/ClientStreamFactory.java:
```diff
@@ -1401,30 +1440,18 @@ private void doReset(
-                throw ex;
-            }
-        }
+        tlsEngine.closeInbound();
```
Remove `doCloseInbound(...)` and replace with direct calls to `tlsEngine.closeInbound()`.

jitsni replied:
done

jfallows commented on src/main/java/org/reaktivity/nukleus/tls/internal/stream/ServerStreamFactory.java:
```diff
@@ -1301,6 +1358,10 @@ private void flushNetwork(
+            if (applicationReplyCleaner != null)
+            {
+                applicationReplyCleaner.run();      // sends RESET to application reply stream (if not received END)
+            }
```
Instead of using `null`, lets make the `applicationReplyCleaner` (renamed to `networkReplyDoneHandler`) a no-op lambda when necessary.

jfallows commented on src/main/java/org/reaktivity/nukleus/tls/internal/stream/ServerStreamFactory.java:
```diff
@@ -1281,16 +1326,28 @@ private void handleWindow(
-        long networkReplyId)
+        long networkReplyId,
+        Runnable applicationReplyCleaner)
     {
```
Rename `applicationReplyCleaner` to `networkReplyDoneHandler`

jfallows commented on src/main/java/org/reaktivity/nukleus/tls/internal/stream/ServerStreamFactory.java:
```diff
@@ -242,6 +242,11 @@ private RouteFW wrapRoute(
+
+        private long networkCorrelationId;
+
+        private Runnable applicationReplyCleaner;
```
Rename `applicationReplyCleaner` to `networkReplyDoneHandler`.

jitsni replied:
done

jfallows commented on src/main/java/org/reaktivity/nukleus/tls/internal/stream/ServerStreamFactory.java:
```diff
@@ -1441,30 +1502,19 @@ private void doReset(
         MessageConsumer networkReply,
-        long networkReplyId) throws SSLException
+        long networkReplyId,
+        Runnable applicationReplyCleaner) throws SSLException
```
Rename `applicationReplyCleaner` to `networkReplyDoneHandler`.

jfallows commented on src/main/java/org/reaktivity/nukleus/tls/internal/stream/ServerStreamFactory.java:
```diff
@@ -1281,16 +1326,28 @@ private void handleWindow(
+            {
+                doReset(applicationReplyThrottle, applicationReplyId);
+            }
+        }
```
Rename `doResetApplicationReplyStream` to `handleNetworkReplyDone`.

jitsni replied:
done

jfallows commented on src/main/java/org/reaktivity/nukleus/tls/internal/stream/ServerStreamFactory.java:
```diff
@@ -766,14 +778,38 @@ private void handleReset(
+            Runnable applicationReplyCleaner)
+        {
+            this.applicationReplyCleaner = applicationReplyCleaner;
+        }
```
Rename as `setNetworkReplyDoneHandler` method and `networkReplyDoneHandler` field and parameter.

jitsni replied:
done

jfallows commented on src/main/java/org/reaktivity/nukleus/tls/internal/stream/ClientStreamFactory.java:
```diff
@@ -1250,7 +1287,8 @@ private void flushNetwork(
         MessageConsumer networkTarget,
-        long networkId)
+        long networkId,
+        Runnable applicationCleaner)
```
Rename `applicationCleaner` to `networkReplyDoneHandler`.

jitsni replied:
done

jfallows commented on src/main/java/org/reaktivity/nukleus/tls/internal/stream/ClientStreamFactory.java:
```diff
@@ -1401,30 +1440,18 @@ private void doReset(
         MessageConsumer networkTarget,
-        long networkId) throws SSLException
+        long networkId,
+        Runnable applicationCleaner) throws SSLException
```
Rename `applicationCleaner` to `networkReplyDoneHandler`.

jitsni replied:
done

jfallows commented on src/main/java/org/reaktivity/nukleus/tls/internal/stream/ClientStreamFactory.java:
```diff
@@ -505,16 +524,24 @@ private ClientHandshake(
             this.windowHandler = this::beforeNetworkReply;
+            this.applicationThrottle = applicationThrottle;
+            this.applicationId = applicationId;
+            this.applicationCleaner = applicationCleaner;
```
Rename `applicationCleaner` to `networkOutboundDoneHandler`.

jitsni replied:
done

jfallows commented on src/main/java/org/reaktivity/nukleus/tls/internal/stream/ClientStreamFactory.java:
```diff
@@ -471,6 +476,14 @@ private void handleReset(
+            {
+                doReset(applicationThrottle, applicationId);
+            }
+        }
```
Rename `doResetApplicationStream` to `handleNetworkReplyDone`.

jitsni replied:
done

jfallows commented on src/main/java/org/reaktivity/nukleus/tls/internal/stream/ClientStreamFactory.java:
```diff
@@ -603,16 +630,16 @@ private void handleReset(
             finally
             {
-                networkThrottle.accept(reset.typeId(), reset.buffer(), reset.offset(), reset.sizeof());
+                doReset(applicationThrottle, applicationId);
```
The `reaktor` will implicitly `RESET` if we successfully remove the correlation, so this seems like it could trigger a duplicate `RESET` for that case.

Perhaps change the `try` block as follows:
```
if (correlations.remove(networkCorrelationId) == null)
{
    doReset(applicationThrottle, applicationId);
}
doCloseInbound(tlsEngine);
```
...and remove the `finally` block.

jfallows commented on src/main/java/org/reaktivity/nukleus/tls/internal/stream/ServerStreamFactory.java:
```diff
@@ -862,7 +903,7 @@ private void handleData(
                 if (networkSlot == NO_SLOT)
                 {
-                    doCloseOutbound(tlsEngine, networkReply, networkReplyId);
+                    doCloseOutbound(tlsEngine, networkReply, networkReplyId, null);
```
Pass a no-op lambda instead of `null`.

jfallows commented on src/main/java/org/reaktivity/nukleus/tls/internal/stream/ServerStreamFactory.java:
```diff
@@ -943,7 +984,7 @@ private void handleEnd(
             try
             {
-                doCloseOutbound(tlsEngine, networkReply, networkReplyId);
+                doCloseOutbound(tlsEngine, networkReply, networkReplyId, null);
```
Pass a no-op lambda instead of `null`.

jfallows commented on src/main/java/org/reaktivity/nukleus/tls/internal/stream/ServerStreamFactory.java:
```diff
@@ -955,9 +996,10 @@ private void handleEnd(
-            doAbort(networkReply, networkReplyId);
+
+            // Sends ABORT to application
+            applicationCleaner.run();
```
Rename `applicationCleaner` to `networkReplyDoneHandler`.

jfallows commented on src/main/java/org/reaktivity/nukleus/tls/internal/stream/ServerStreamFactory.java:
```diff
@@ -974,17 +1016,16 @@ private void updateNetworkWindow(
+        private void setApplicationReplyCleaner(
+            Runnable applicationReplyCleaner)
+        {
+            applicationInCleanerConsumer.accept(applicationReplyCleaner);
```
Rename `applicationReplyCleaner` to `networkReplyDoneHandler`.
Rename `applicationInCleaner` to `networkReplyDoneHandlerConsumer`.
Rename `setApplicationReplyCleaner` to `setNetworkReplyDoneHandler`.

jitsni committed:
- Minor renmaing
- Using NOP for cleaner runnables instead of null
- Making sure handleReset doesn't send duplicate RESET

jitsni committed:
One more minor renaming of field
